### PR TITLE
Migrates repo to use Bicep Module Registry conventions

### LINKED ISSUE #5: CI/CD pipeline for managing published modules
This should include the following considerations:
1. Versioning
2. Code analysis & Linting
2. Testing
3. Publishing of pre-release artefacts
4. Publishing of final release artefacts

### PR DESCRIPTION
**Pull request by JamesDawson**
* Restructures repo layout to align with the conventions defined for the public Bicep module registry
* Adds build process using `Endjin.RecommendedPractices.Build`
* Adds a build workflow to manage the module publishing policy

Resolves #5

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    - uses: actions/setup-dotnet@v1
+      if: ${{ github.actor != 'nektos/act' }}
+      name: Install .NET Core SDK 3.1.x
+      with:
+        dotnet-version: '3.1.x'
+    - uses: actions/setup-dotnet@v1
+      if: ${{ github.actor != 'nektos/act' }}
+      name: Install .NET Core SDK 5.0.x
+      with:
+        dotnet-version: '5.0.x'
```
Comment (HowardvanRooijen) [inline .github/workflows/build.yml:28]: Do we still need .NET 5.0 or should this be 6.0 now?

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (HowardvanRooijen) [inline .vscode/launch.json:1]: Should this file be included or should the `/.vscode/` folder be excluded?

File: modules/general/app-configuration/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+| :-------------------------- | :------: | :------: | :----------------------------------------------------------------------------------------------------------------------------------------- |
+| `name`                      | `string` | Yes      | The name of the app configuration store                                                                                                    |
+| `location`                  | `string` | Yes      | The location of the app configuration store                                                                                                |
+| `enablePublicNetworkAccess` | `bool`   | No       | When false, the app configuration store will be inaccessible via its public IP address                                                     |
+| `sku`                       | `string` | No       | SKU for the app configuration store                                                                                                        |
+| `useExisting`               | `bool`   | No       | When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated |
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/README.md:13]: typo "udpated"

File: modules/general/acr/test/main.test.json [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, admin access via the ACR key is enabled; When false, access is via RBAC"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing ACR will be returned; When false, the ACR is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/acr/test/main.test.json:87]: Typo "udpated"

File: modules/general/app-configuration/main.bicep [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+param enablePublicNetworkAccess bool = true
+
+@description('SKU for the app configuration store')
+@allowed([
+  'Free'
+  'Standard'
+])
+param sku string = 'Standard'
+
+@description('When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated')
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.bicep:21]: Typo "udpated"

File: modules/general/app-configuration/main.bicep [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+param useExisting bool = false
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
+
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.bicep:27]: extra line

File: modules/general/app-configuration/main.bicep [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+var publicNetworkAccess = enablePublicNetworkAccess ? 'Enabled' : 'Disabled'
+
+
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.bicep:30]: extra line

File: modules/general/app-configuration/main.bicep [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.bicep:32]: Extra line

File: modules/general/app-configuration/main.json [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      ],
+      "metadata": {
+        "description": "SKU for the app configuration store"
+      }
+    },
+    "useExisting": {
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.json:46]: typo "udpated"

File: modules/general/app-configuration/test/main.test.json [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+              ],
+              "metadata": {
+                "description": "SKU for the app configuration store"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/test/main.test.json:84]: Typo "Udpated"

File: modules/general/app-insights-with-config-publish/README.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+| `keyVaultName`              | `string` | Yes      | The key vault name where the instrumentation key will be stored                                                                            |
+| `keyVaultResourceGroupName` | `string` | Yes      | The resource group of the key vault where the instrumentation key will be stored                                                           |
+| `keyVaultSubscriptionId`    | `string` | No       | The subscription of the key vault where the instrumentation key will be stored                                                             |
+| `kind`                      | `string` | No       | The kind of application using the workspace                                                                                                |
+| `applicationType`           | `string` | No       | The type of application using the workspace                                                                                                |
+| `useExisting`               | `bool`   | No       | When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated |
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/README.md:17]: Typo "Udpated"

File: modules/general/app-insights-with-config-publish/main.bicep [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+param kind string = 'web'
+
+@description('The type of application using the workspace')
+@allowed([
+  'other'
+  'web'
+])
+param applicationType string = 'web'
+
+@description('When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated')
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/main.bicep:41]: Typo "Udpated"

File: modules/general/app-insights-with-config-publish/main.bicep [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+param useExisting bool = false
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/main.bicep:46]: Extra line

File: modules/general/app-insights-with-config-publish/main.bicep [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/main.bicep:49]: Extra Line

File: modules/general/app-insights-with-config-publish/main.json [hunk 15]
```diff
@@ -0,0 +1,346 @@
...
+      ],
+      "metadata": {
+        "description": "The type of application using the workspace"
+      }
+    },
+    "useExisting": {
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/main.json:80]: Typo "Udpated"

File: modules/general/app-insights-with-config-publish/test/main.test.json [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+              ],
+              "metadata": {
+                "description": "The type of application using the workspace"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/test/main.test.json:137]: Typo "Udpated"

File: modules/general/app-insights-with-config-publish/test/main.test.json [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+                      ],
+                      "metadata": {
+                        "description": "The type of application using the workspace"
+                      }
+                    },
+                    "useExisting": {
+                      "type": "bool",
+                      "defaultValue": false,
+                      "metadata": {
+                        "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/test/main.test.json:231]: Typo "Udpated"

File: modules/general/app-insights/README.md [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+| :---------------- | :------: | :------: | :----------------------------------------------------------------------------------------------------------------------------------------- |
+| `name`            | `string` | Yes      | The name of the app insights workspace                                                                                                     |
+| `location`        | `string` | Yes      | The location of the app insights workspace                                                                                                 |
+| `kind`            | `string` | No       | The kind of application using the workspace                                                                                                |
+| `applicationType` | `string` | No       | The type of application using the workspace                                                                                                |
+| `useExisting`     | `bool`   | No       | When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated |
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/README.md:13]: Typo "Udpated"

File: modules/general/app-insights/main.bicep [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+param kind string = 'web'
+
+@description('The type of application using the workspace')
+@allowed([
+  'other'
+  'web'
+])
+param applicationType string = 'web'
+
+@description('When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated')
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.bicep:29]: Typo "Udpated"

File: modules/general/app-insights/main.bicep [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+param useExisting bool = false
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.bicep:34]: Extra line

File: modules/general/app-insights/main.bicep [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+targetScope = 'resourceGroup'
+
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.bicep:38]: Extra line

File: modules/general/app-insights/main.bicep [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+resource app_insights 'Microsoft.Insights/components@2020-02-02' = if (!useExisting) {
+  name: name
+  location: location
+  kind: kind
+  properties: {
+    Application_Type: applicationType
+  }
+  tags: resourceTags
+}
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.bicep:54]: Extra line

File: modules/general/app-insights/main.bicep [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+
+@description('The resource ID of the app insights workspace')
+output id string = useExisting ? existing_app_insights.id : app_insights.id
+@description('The name of the app insights workspace')
+output name string =  useExisting ? existing_app_insights.name : app_insights.name
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.bicep:60]: Extra line

File: modules/general/app-insights/main.json [hunk 24]
```diff
@@ -0,0 +1,102 @@
...
+      ],
+      "metadata": {
+        "description": "The type of application using the workspace"
+      }
+    },
+    "useExisting": {
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/main.json:54]: Typo "Udpated"

File: modules/general/app-insights/test/main.test.json [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+              ],
+              "metadata": {
+                "description": "The type of application using the workspace"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing app configuration store will be returned; When false, the app configuration store is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-insights/test/main.test.json:92]: Typo "Udpated"

File: modules/general/app-service-plan/README.md [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+| `targetWorkerSizeId`                 | `int`    | No       | The target size of the workers                                                                                               |
+| `elasticScaleEnabled`                | `bool`   | No       | When true, elastic scale is enabled                                                                                          |
+| `maximumElasticWorkerCount`          | `int`    | No       | The maximum number of elastic workers                                                                                        |
+| `zoneRedundant`                      | `bool`   | No       | When true, the app service will have zone redundancy                                                                         |
+| `resourceTags`                       | `object` | No       | The resource tags applied to resources                                                                                       |
+| `useExisting`                        | `bool`   | No       | When true, the details of an existing app service plan will be returned; When false, the app service plan is created/udpated |
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/README.md:19]: Typo "Udpated"

File: modules/general/app-service-plan/main.bicep [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+
+@description('When true, the details of an existing app service plan will be returned; When false, the app service plan is created/udpated')
+param useExisting bool = false
+
+@description('The resource group in which the existing app service plan resides')
+param existingPlanResourceGroupName string = resourceGroup().name
+
+@description('The subscription in which the existing app service plan resides')
+param existingPlanResourceSubscriptionId string = subscription().subscriptionId
+
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/main.bicep:43]: Extra line

File: modules/general/app-service-plan/main.bicep [hunk 28]
```diff
@@ -0,0 +1,77 @@
...
+
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/main.bicep:46]: Extra line

File: modules/general/app-service-plan/main.bicep [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+    zoneRedundant: zoneRedundant
+  }
+  tags: resourceTags
+}
+
+@description('The resource ID of the app service plan')
+output id string = useExisting ? existing_hosting_plan.id : hosting_plan.id
+@description('The name of the app service plan')
+output nmae string = useExisting ? existing_hosting_plan.name : hosting_plan.name
+
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/main.bicep:75]: Extra line

File: modules/general/app-service-plan/main.bicep [hunk 30]
```diff
@@ -0,0 +1,77 @@
...
+    targetWorkerSizeId: targetWorkerSizeId
+    zoneRedundant: zoneRedundant
+  }
+  tags: resourceTags
+}
+
+@description('The resource ID of the app service plan')
+output id string = useExisting ? existing_hosting_plan.id : hosting_plan.id
+@description('The name of the app service plan')
+output nmae string = useExisting ? existing_hosting_plan.name : hosting_plan.name
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/main.bicep:74]: Typo "nmae"

File: modules/general/app-service-plan/main.json [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+      "defaultValue": {},
+      "metadata": {
+        "description": "The resource tags applied to resources"
+      }
+    },
+    "useExisting": {
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "When true, the details of an existing app service plan will be returned; When false, the app service plan is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/main.json:82]: Typo "Udpated"

File: modules/general/app-service-plan/test/main.test.json [hunk 32]
```diff
@@ -0,0 +1,193 @@
...
+              "defaultValue": {},
+              "metadata": {
+                "description": "The resource tags applied to resources"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing app service plan will be returned; When false, the app service plan is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/app-service-plan/test/main.test.json:126]: Typo "Udpated"

File: modules/general/container-app-environment-with-config-publish/main.bicep [hunk 33]
```diff
@@ -0,0 +1,101 @@
...
+
+@description('The name of the key vault where the app insights instrumentation key will be published')
+param keyVaultName string
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
+
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment-with-config-publish/main.bicep:43]: Extra Line

File: modules/general/container-app-environment-with-config-publish/main.bicep [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+@description('The app configuration label to apply to the published config')
+param appConfigurationLabel string
+
+@description('The name of the key vault where the app insights instrumentation key will be published')
+param keyVaultName string
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment-with-config-publish/main.bicep:40]: Extra Line

File: modules/general/container-app-environment-with-config-publish/main.bicep [hunk 35]
```diff
@@ -0,0 +1,101 @@
...
+module aikey_secret '../key-vault-secret/main.bicep' = {
+  name: 'aiKeySecretDeploy'
+  params: {
+    keyVaultName: keyVaultName
+    secretName: 'AppInsightsInstrumentationKey'
+    contentValue: app_environment.outputs.appinsights_instrumentation_key
+    contentType: 'text/plain'
+  }
+}
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment-with-config-publish/main.bicep:85]: Extra Line

File: modules/general/container-app-environment-with-config-publish/test/main.test.json [hunk 36]
```diff
@@ -0,0 +1,765 @@
...
+                              "defaultValue": false,
+                              "metadata": {
+                                "description": "When true, admin access via the ACR key is enabled; When false, access is via RBAC"
+                              }
+                            },
+                            "useExisting": {
+                              "type": "bool",
+                              "defaultValue": false,
+                              "metadata": {
+                                "description": "When true, the details of an existing ACR will be returned; When false, the ACR is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment-with-config-publish/test/main.test.json:428]: Typo "Udpated"

File: modules/general/container-app-environment-with-config-publish/test/main.test.json [hunk 37]
```diff
@@ -0,0 +1,765 @@
...
+                        "value": "[parameters('contentValue')]"
+                      }
+                    }
+                  ],
+                  "outputs": {
+                    "secretUriWithVersion": {
+                      "type": "string",
+                      "value": "[if(parameters('useExisting'), reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion, reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion)]",
+                      "metadata": {
+                        "description": "The key vault URI linking to the new/udpated secret"
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment-with-config-publish/test/main.test.json:706]: Typo "Udpated"

File: modules/general/container-app-environment/main.bicep [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+
+@description('The SKU for the container registry')
+param containerRegistrySku string = 'Standard'
+
+@description('When true, admin access via the ACR key is enabled; When false, access is via RBAC')
+param enableContainerRegistryAdminUser bool = true
+
+@description('The resource tags applied to resources')
+param resourceTags object = {}
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment/main.bicep:46]: Extra line

File: modules/general/container-app-environment/main.bicep [hunk 39]
```diff
@@ -0,0 +1,145 @@
...
+
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment/main.bicep:49]: Extra line

File: modules/general/container-app-environment/main.bicep [hunk 40]
```diff
@@ -0,0 +1,145 @@
...
+  name: 'appEnvAcrDeploy'
+  params: {
+    name: containerRegistryName
+    location: location
+    sku: containerRegistrySku
+    adminUserEnabled: enableContainerRegistryAdminUser
+    resourceTags: resourceTags
+  }
+}
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment/main.bicep:61]: Extra line

File: modules/general/container-app-environment/main.bicep [hunk 41]
```diff
@@ -0,0 +1,145 @@
...
+  properties: {
+    Application_Type: 'web'
+    #disable-next-line BCP036
+    Flow_Type: 'Redfield'
+    #disable-next-line BCP036
+    Request_Source: 'CustomDeployment'
+  }
+  tags: resourceTags
+}
+
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment/main.bicep:98]: Extra line

File: modules/general/container-app-environment/main.json [hunk 42]
```diff
@@ -0,0 +1,351 @@
...
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, admin access via the ACR key is enabled; When false, access is via RBAC"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing ACR will be returned; When false, the ACR is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/container-app-environment/main.json:239]: Typo "Udpated"

File: modules/general/key-vault-secret/README.md [hunk 43]
```diff
@@ -0,0 +1,31 @@
...
+| `contentType`  | `string`       | No       | Type of the secret                                |
+| `contentValue` | `secureString` | No       | Value of the secret                               |
+| `keyVaultName` | `string`       | Yes      | Name of the vault                                 |
+| `useExisting`  | `bool`         | No       | When true, a pre-existing secret will be returned |
+
+## Outputs
+
+| Name                 | Type   | Description                                         |
+| :------------------- | :----: | :-------------------------------------------------- |
+| secretUriWithVersion | string | The key vault URI linking to the new/udpated secret |
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/README.md:19]: Typo "Udpated"

File: modules/general/key-vault-secret/main.bicep [hunk 44]
```diff
@@ -0,0 +1,44 @@
...
+@description('Value of the secret')
+@secure()
+param contentValue string = ''
+
+@description('Name of the vault')
+param keyVaultName string
+
+@description('When true, a pre-existing secret will be returned')
+param useExisting bool = false
+
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/main.bicep:20]: Extra line

File: modules/general/key-vault-secret/main.bicep [hunk 45]
```diff
@@ -0,0 +1,44 @@
...
+
+targetScope = 'resourceGroup'
+
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/main.bicep:23]: Extra line

File: modules/general/key-vault-secret/main.bicep [hunk 46]
```diff
@@ -0,0 +1,44 @@
...
+resource secret 'Microsoft.KeyVault/vaults/secrets@2021-06-01-preview' = if (!useExisting) {
+  name: secretName
+  parent: key_vault
+  properties: {
+    contentType: contentType
+    value: contentValue
+  }
+}
+
+@description('The key vault URI linking to the new/udpated secret')
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/main.bicep:43]: Typo "Udpated"

File: modules/general/key-vault-secret/main.json [hunk 47]
```diff
@@ -0,0 +1,67 @@
...
+        "value": "[parameters('contentValue')]"
+      }
+    }
+  ],
+  "outputs": {
+    "secretUriWithVersion": {
+      "type": "string",
+      "value": "[if(parameters('useExisting'), reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion, reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion)]",
+      "metadata": {
+        "description": "The key vault URI linking to the new/udpated secret"
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/main.json:63]: Typo "Udpated"

File: modules/general/key-vault-secret/test/main.test.json [hunk 48]
```diff
@@ -0,0 +1,125 @@
...
+                "value": "[parameters('contentValue')]"
+              }
+            }
+          ],
+          "outputs": {
+            "secretUriWithVersion": {
+              "type": "string",
+              "value": "[if(parameters('useExisting'), reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion, reference(resourceId('Microsoft.KeyVault/vaults/secrets', parameters('keyVaultName'), parameters('secretName')), '2021-06-01-preview').secretUriWithVersion)]",
+              "metadata": {
+                "description": "The key vault URI linking to the new/udpated secret"
```
Comment (HowardvanRooijen) [inline modules/general/key-vault-secret/test/main.test.json:114]: Typo "Udpated"

File: modules/general/key-vault/README.md [hunk 49]
```diff
@@ -0,0 +1,42 @@
...
+| `enabledForTemplateDeployment`  | `bool`   | No       | When true, the key vault will be accessible by ARM deployments                                                             |
+| `enableDiagnostics`             | `bool`   | Yes      | When true, diagnostics settings will be enabled for the key vault                                                          |
+| `diagnosticsStorageAccountName` | `string` | No       | The storage account name to be used for key vault diagnostic settings                                                      |
+| `useExistingStorageAccount`     | `bool`   | No       | When true, an existing storage account be used for diagnotics settings; When false, the storage account is created/updated |
+| `diagnosticsRetentionDays`      | `int`    | No       | Sets the retention policy for diagnostics settings data, in days                                                           |
+| `useExisting`                   | `bool`   | No       | When true, the details of an existing key vault will be returned; When false, the key vault is created/udpated             |
```
Comment (HowardvanRooijen) [inline modules/general/key-vault/README.md:21]: Typo "Udpated"

File: modules/general/key-vault/main.bicep [hunk 50]
```diff
@@ -0,0 +1,110 @@
...
+    ]
+  }
+}
+
+
+@description('The resource ID of the key vault')
+output id string = useExisting ? existing_key_vault.id : key_vault.id
+@description('The name of the key vault')
+output name string = useExisting ? existing_key_vault.name : key_vault.name
+
```
Comment (HowardvanRooijen) [inline modules/general/key-vault/main.bicep:108]: Is this an extra line, or as in other files you break the outputs up by grouping them with a comment?

File: modules/general/key-vault/test/main.test.json [hunk 51]
```diff
@@ -0,0 +1,490 @@
...
+              "defaultValue": 30,
+              "metadata": {
+                "description": "Sets the retention policy for diagnostics settings data, in days"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing key vault will be returned; When false, the key vault is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/key-vault/test/main.test.json:140]: Typo "Udpated"

File: modules/general/resource-group/README.md [hunk 52]
```diff
@@ -0,0 +1,32 @@
...
+
+Deploys a resource group or gets a reference to an existing one - useful for deferred evaluation of target resource group details
+
+## Parameters
+
+| Name           | Type     | Required | Description                                                                                                              |
+| :------------- | :------: | :------: | :----------------------------------------------------------------------------------------------------------------------- |
+| `name`         | `string` | Yes      | The name of the resource group                                                                                           |
+| `location`     | `string` | Yes      | The location of the resource group                                                                                       |
+| `useExisting`  | `bool`   | No       | When true, the details of an existing resource group will be returned; When false, the resource group is created/udpated |
```
Comment (HowardvanRooijen) [inline modules/general/resource-group/README.md:11]: Typo "Udpated"

File: modules/general/resource-group/test/main.test.json [hunk 53]
```diff
@@ -0,0 +1,115 @@
...
+              "type": "string",
+              "metadata": {
+                "description": "The location of the resource group"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing resource group will be returned; When false, the resource group is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/resource-group/test/main.test.json:67]: Typo "Udpated"

File: modules/general/storage-account/main.bicep [hunk 54]
```diff
@@ -0,0 +1,91 @@
...
+    contentValue: useExisting ? existing_storage_account.listKeys().keys[0].value : storage_account.listKeys().keys[0].value
+  }
+}
+
+
+@description('The resource ID of the storage account')
+output id string = useExisting ? existing_storage_account.id : storage_account.id
+@description('The name of the storage account')
+output name string = useExisting ? existing_storage_account.name : storage_account.name
+
```
Comment (HowardvanRooijen) [inline modules/general/storage-account/main.bicep:89]: Extra line, or grouping without comment?

File: modules/general/storage-account/main.json [hunk 55]
```diff
@@ -0,0 +1,247 @@
...
+      "defaultValue": "",
+      "metadata": {
+        "description": "The key vault secret name used to store the access key"
+      }
+    },
+    "useExisting": {
+      "type": "bool",
+      "defaultValue": false,
+      "metadata": {
+        "description": "When true, the details of an existing storage account will be returned; When false, the storage account is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/storage-account/main.json:99]: Typo "Udpated"

File: modules/general/storage-account/test/main.test.json [hunk 56]
```diff
@@ -0,0 +1,289 @@
...
+              "defaultValue": "",
+              "metadata": {
+                "description": "The key vault secret name used to store the access key"
+              }
+            },
+            "useExisting": {
+              "type": "bool",
+              "defaultValue": false,
+              "metadata": {
+                "description": "When true, the details of an existing storage account will be returned; When false, the storage account is created/udpated"
```
Comment (HowardvanRooijen) [inline modules/general/storage-account/test/main.test.json:137]: Typo "Udpated"

File: tests/integration.tests.ps1 [hunk 57]
```diff
@@ -0,0 +1,7 @@
+Describe "Bicep Module Integration Tests" {
+    Context "Dummy Module" {
+        It "should be true" {
+            $true | SHould -Be $true
```
Comment (HowardvanRooijen) [inline tests/integration.tests.ps1:4]: Typo "SHould"

File: modules/general/acr/main.bicep [hunk 58]
```diff
@@ -0,0 +1,56 @@
...
+  tags: resourceTags
+}
+
+@description('The resource ID of the container registry')
+output id string = useExisting ? existing_acr.id : acr.id
+@description('The name of the container registry')
+output name string = useExisting ? existing_acr.name : acr.name
+@description('The admin username of the container registry')
+output loginServer string = useExisting ? existing_acr.properties.loginServer : acr.properties.loginServer
+
```
Comment (HowardvanRooijen) [inline modules/general/acr/main.bicep:54]: Extra line or grouping without a comment?

File: modules/general/app-configuration/main.bicep [hunk 59]
```diff
@@ -0,0 +1,56 @@
...
+    publicNetworkAccess: publicNetworkAccess
+  }
+  tags: resourceTags
+}
+
+@description('The resource ID of the app configuration store')
+output id string = useExisting ? existing_app_config_store.id : app_config_store.id
+@description('The name of the app configuration store')
+output name string = useExisting ? existing_app_config_store.name : app_config_store.name
+
```
Comment (HowardvanRooijen) [inline modules/general/app-configuration/main.bicep:54]: Extra line or grouping without a comment?

File: modules/general/app-insights-with-config-publish/main.bicep [hunk 60]
```diff
@@ -0,0 +1,80 @@
...
+    contentValue: app_insights.outputs.appInsightsWorkspaceResource.properties.InstrumentationKey
+    contentType: 'text/plain'
+  }
+}
+
+@description('The resource ID of the app insights workspace')
+output id string = app_insights.outputs.id
+@description('The name of the app insights workspace')
+output name string = app_insights.outputs.name
+
```
Comment (HowardvanRooijen) [inline modules/general/app-insights-with-config-publish/main.bicep:78]: extra line or grouping without a comment?

File: .vscode/launch.json (hunk 2 above)
Comment (MikeLarah) [reply .vscode/launch.json:1]: I think it's generally good practice to include the .vscode folder so everyone working on the project has the same setup

File: modules/general/acr/README.md [hunk 61]
```diff
@@ -0,0 +1,39 @@
+# container-registry
+
+Azure Container Registry
+
+## Description
+
+{{ Add detailed description for the module. }}
```
Comment (MikeLarah) [inline modules/general/acr/README.md:7]: Is this now auto-generated as placeholder text?

File: tests/integration.tests.ps1 [hunk 62]
```diff
@@ -0,0 +1,7 @@
+Describe "Bicep Module Integration Tests" {
+    Context "Dummy Module" {
+        It "should be true" {
+            $true | should -Be $true
+        }
+    }
+}
```
Comment (MikeLarah) [inline tests/integration.tests.ps1:7]: What's the purpose of this?

Comment (JamesDawson) [reply tests/integration.tests.ps1:7]: It's just some scaffolding for when we want to run the module tests are part of the build - I'll add a comment.

File: modules/general/acr/README.md (hunk 61 above)
Comment (JamesDawson) [reply modules/general/acr/README.md:7]: There are bunch of those across the modules resulting from the updates to the `brm` tooling - I was thinking we can address those in a follow-up PR, which I've now added a task for: #14

**Review by MikeLarah — APPROVED** [review]

**Pull request merged**